File: vm/src/main/java/com/libreshockwave/vm/opcode/PropertyIdMappings.java
package com.libreshockwave.vm.opcode;

import java.util.Map;

/**
 * Mappings from property IDs to property names for GET/SET opcodes.
 * Based on dirplayer-rs constants.rs
 */
public final class PropertyIdMappings {

    private PropertyIdMappings() {}

    /**
     * Movie property names (property type 0x00).
     * Used for "the floatPrecision", "the mouseDownScript", etc.
     */
    public static final Map<Integer, String> MOVIE_PROPS = Map.ofEntries(
        Map.entry(0x00, "floatPrecision"),
        Map.entry(0x01, "mouseDownScript"),
        Map.entry(0x02, "mouseUpScript"),
        Map.entry(0x03, "keyDownScript"),
        Map.entry(0x04, "keyUpScript"),
        Map.entry(0x05, "timeoutScript"),
        Map.entry(0x06, "short time"),
        Map.entry(0x07, "abbr time"),
        Map.entry(0x08, "long time"),
        Map.entry(0x09, "short date"),
        Map.entry(0x0a, "abbr date"),
        Map.entry(0x0b, "long date")
    );

    /**
     * Sprite property names (property type 0x06).
     */
    public static final Map<Integer, String> SPRITE_PROPS = Map.ofEntries(
        Map.entry(0x01, "type"),
        Map.entry(0x02, "backColor"),
        Map.entry(0x03, "bottom"),
        Map.entry(0x04, "castNum"),
        Map.entry(0x05, "constraint"),
        Map.entry(0x06, "cursor"),
        Map.entry(0x07, "foreColor"),
        Map.entry(0x08, "height"),
        Map.entry(0x09, "immediate"),
        Map.entry(0x0a, "ink"),
        Map.entry(0x0b, "left"),
        Map.entry(0x0c, "lineSize"),
        Map.entry(0x0d, "locH"),
        Map.entry(0x0e, "locV"),
        Map.entry(0x0f, "movieRate"),
        Map.entry(0x10, "movieTime"),
        Map.entry(0x11, "pattern"),
        Map.entry(0x12, "puppet"),
        Map.entry(0x13, "right"),
        Map.entry(0x14, "startTime"),
        Map.entry(0x15, "stopTime"),
        Map.entry(0x16, "stretch"),
        Map.entry(0x17, "top"),
        Map.entry(0x18, "trails"),
        Map.entry(0x19, "visible"),
        Map.entry(0x1a, "volume"),
        Map.entry(0x1b, "width"),
        Map.entry(0x1c, "blend"),
        Map.entry(0x1d, "scriptNum"),
        Map.entry(0x1e, "moveableSprite"),
        Map.entry(0x1f, "editableText"),
        Map.entry(0x20, "scoreColor"),
        Map.entry(0x21, "loc"),
        Map.entry(0x22, "rect"),
        Map.entry(0x23, "memberNum"),
        Map.entry(0x24, "castLibNum"),
        Map.entry(0x25, "member"),
        Map.entry(0x26, "scriptInstanceList"),
        Map.entry(0x27, "currentTime"),
        Map.entry(0x28, "mostRecentCuePoint"),
        Map.entry(0x29, "tweened"),
        Map.entry(0x2a, "name")
    );

    /**
     * Animation property names (property type 0x07).
     */
    public static final Map<Integer, String> ANIM_PROPS = Map.ofEntries(
        Map.entry(0x01, "beepOn"),
        Map.entry(0x02, "buttonStyle"),
        Map.entry(0x03, "centerStage"),
        Map.entry(0x04, "checkBoxAccess"),
        Map.entry(0x05, "checkboxType"),
        Map.entry(0x06, "colorDepth"),
        Map.entry(0x07, "colorQD"),
        Map.entry(0x08, "exitLock"),
        Map.entry(0x09, "fixStageSize"),
        Map.entry(0x0a, "fullColorPermit"),
        Map.entry(0x0b, "imageDirect"),
        Map.entry(0x0c, "doubleClick"),
        Map.entry(0x0d, "key"),
        Map.entry(0x0e, "lastClick"),
        Map.entry(0x0f, "lastEvent"),
        Map.entry(0x10, "keyCode"),
        Map.entry(0x11, "lastKey"),
        Map.entry(0x12, "lastRoll"),
        Map.entry(0x13, "timeoutLapsed"),
        Map.entry(0x14, "multiSound"),
        Map.entry(0x15, "pauseState"),
        Map.entry(0x16, "quickTimePresent"),
        Map.entry(0x17, "selEnd"),
        Map.entry(0x18, "selStart"),
        Map.entry(0x19, "soundEnabled"),
        Map.entry(0x1a, "soundLevel"),
        Map.entry(0x1b, "stageColor"),
        Map.entry(0x1d, "switchColorDepth"),
        Map.entry(0x1e, "timeoutKeyDown"),
        Map.entry(0x1f, "timeoutLength"),
        Map.entry(0x20, "timeoutMouse"),
        Map.entry(0x21, "timeoutPlay"),
        Map.entry(0x22, "timer"),
        Map.entry(0x23, "preLoadRAM"),
        Map.entry(0x24, "videoForWindowsPresent"),
        Map.entry(0x25, "netPresent"),
        Map.entry(0x26, "safePlayer"),
        Map.entry(0x27, "soundKeepDevice"),
        Map.entry(0x28, "soundMixMedia")
    );

    /**
     * Animation2 property names (property type 0x08).
     */
    public static final Map<Integer, String> ANIM2_PROPS = Map.of(
        0x01, "perFrameHook",
        0x02, "number of castMembers",
        0x03, "number of menus",
        0x04, "number of castLibs",
        0x05, "number of xtras"
    );

    /**
     * Sound channel property names (property type 0x04).
     */
    public static final Map<Integer, String> SOUND_PROPS = Map.of(
        0x01, "volume",
        0x02, "pan",
        0x03, "loopCount",
        0x04, "startTime",
        0x05, "endTime",
        0x06, "loopStartTime",
        0x07, "loopEndTime"
    );

    public static String getMoviePropName(int id) {
        return MOVIE_PROPS.get(id);
    }

    public static String getSpritePropName(int id) {
        return SPRITE_PROPS.get(id);
    }

    public static String getAnimPropName(int id) {
        return ANIM_PROPS.get(id);
    }

    public static String getAnim2PropName(int id) {
        return ANIM2_PROPS.get(id);
    }

    public static String getSoundPropName(int id) {
        return SOUND_PROPS.getOrDefault(id, "unknown_sound_prop_" + id);
    }
}

File: cpp/src/lingo/vm/PropertyIdMappings.cpp
#include "libreshockwave/lingo/vm/PropertyIdMappings.hpp"

namespace libreshockwave::lingo::vm {

std::optional<std::string_view> PropertyIdMappings::getMoviePropName(int id) {
    switch (id) {
        case 0x00: return "floatPrecision";
        case 0x01: return "mouseDownScript";
        case 0x02: return "mouseUpScript";
        case 0x03: return "keyDownScript";
        case 0x04: return "keyUpScript";
        case 0x05: return "timeoutScript";
        case 0x06: return "short time";
        case 0x07: return "abbr time";
        case 0x08: return "long time";
        case 0x09: return "short date";
        case 0x0A: return "abbr date";
        case 0x0B: return "long date";
        default: return std::nullopt;
    }
}

std::optional<std::string_view> PropertyIdMappings::getSpritePropName(int id) {
    switch (id) {
        case 0x01: return "type";
        case 0x02: return "backColor";
        case 0x03: return "bottom";
        case 0x04: return "castNum";
        case 0x05: return "constraint";
        case 0x06: return "cursor";
        case 0x07: return "foreColor";
        case 0x08: return "height";
        case 0x09: return "immediate";
        case 0x0A: return "ink";
        case 0x0B: return "left";
        case 0x0C: return "lineSize";
        case 0x0D: return "locH";
        case 0x0E: return "locV";
        case 0x0F: return "movieRate";
        case 0x10: return "movieTime";
        case 0x11: return "pattern";
        case 0x12: return "puppet";
        case 0x13: return "right";
        case 0x14: return "startTime";
        case 0x15: return "stopTime";
        case 0x16: return "stretch";
        case 0x17: return "top";
        case 0x18: return "trails";
        case 0x19: return "visible";
        case 0x1A: return "volume";
        case 0x1B: return "width";
        case 0x1C: return "blend";
        case 0x1D: return "scriptNum";
        case 0x1E: return "moveableSprite";
        case 0x1F: return "editableText";
        case 0x20: return "scoreColor";
        case 0x21: return "loc";
        case 0x22: return "rect";
        case 0x23: return "memberNum";
        case 0x24: return "castLibNum";
        case 0x25: return "member";
        case 0x26: return "scriptInstanceList";
        case 0x27: return "currentTime";
        case 0x28: return "mostRecentCuePoint";
        case 0x29: return "tweened";
        case 0x2A: return "name";
        default: return std::nullopt;
    }
}

std::optional<std::string_view> PropertyIdMappings::getAnimPropName(int id) {
    switch (id) {
        case 0x01: return "beepOn";
        case 0x02: return "buttonStyle";
        case 0x03: return "centerStage";
        case 0x04: return "checkBoxAccess";
        case 0x05: return "checkboxType";
        case 0x06: return "colorDepth";
        case 0x07: return "colorQD";
        case 0x08: return "exitLock";
        case 0x09: return "fixStageSize";
        case 0x0A: return "fullColorPermit";
        case 0x0B: return "imageDirect";
        case 0x0C: return "doubleClick";
        case 0x0D: return "key";
        case 0x0E: return "lastClick";
        case 0x0F: return "lastEvent";
        case 0x10: return "keyCode";
        case 0x11: return "lastKey";
        case 0x12: return "lastRoll";
        case 0x13: return "timeoutLapsed";
        case 0x14: return "multiSound";
        case 0x15: return "pauseState";
        case 0x16: return "quickTimePresent";
        case 0x17: return "selEnd";
        case 0x18: return "selStart";
        case 0x19: return "soundEnabled";
        case 0x1A: return "soundLevel";
        case 0x1B: return "stageColor";
        case 0x1D: return "switchColorDepth";
        case 0x1E: return "timeoutKeyDown";
        case 0x1F: return "timeoutLength";
        case 0x20: return "timeoutMouse";
        case 0x21: return "timeoutPlay";
        case 0x22: return "timer";
        case 0x23: return "preLoadRAM";
        case 0x24: return "videoForWindowsPresent";
        case 0x25: return "netPresent";
        case 0x26: return "safePlayer";
        case 0x27: return "soundKeepDevice";
        case 0x28: return "soundMixMedia";
        default: return std::nullopt;
    }
}

std::optional<std::string_view> PropertyIdMappings::getAnim2PropName(int id) {
    switch (id) {
        case 0x01: return "perFrameHook";
        case 0x02: return "number of castMembers";
        case 0x03: return "number of menus";
        case 0x04: return "number of castLibs";
        case 0x05: return "number of xtras";
        default: return std::nullopt;
    }
}

std::string PropertyIdMappings::getSoundPropName(int id) {
    switch (id) {
        case 0x01: return "volume";
        case 0x02: return "pan";
        case 0x03: return "loopCount";
        case 0x04: return "startTime";
        case 0x05: return "endTime";
        case 0x06: return "loopStartTime";
        case 0x07: return "loopEndTime";
        default: return "unknown_sound_prop_" + std::to_string(id);
    }
}

} // namespace libreshockwave::lingo::vm
